File: llm_plugins/huggingface.py
import os
import requests
from .base import LLMPlugin

class HuggingFacePlugin(LLMPlugin):
    def __init__(self):
        self.api_key = os.environ.get("HF_API_KEY")
        self.api_url = "https://api-inference.huggingface.co/models/"
        self._models = os.environ.get("HF_MODELS", "gpt2,distilgpt2").split(",")

    def name(self) -> str:
        return "huggingface"

    def call(self, prompt: str, model: str, max_tokens: int, temperature: float) -> str:
        headers = {"Authorization": f"Bearer {self.api_key}"}
        payload = {
            "inputs": prompt,
            "parameters": {
                "max_new_tokens": max_tokens,
                "temperature": temperature
            },
            "options": {"wait_for_model": True}
        }
        url = self.api_url + model
        resp = requests.post(url, headers=headers, json=payload, timeout=60)
        resp.raise_for_status()
        data = resp.json()
        if isinstance(data, list) and data and "generated_text" in data[0]:
            return data[0]["generated_text"]
        elif isinstance(data, dict) and "error" in data:
            raise RuntimeError(f"HF API error: {data['error']}")
        else:
            return str(data)

    def available_models(self) -> list[str]:
        return [m.strip() for m in self._models if m.strip()]

File: rust/src/plugins/llm/huggingface.rs
use async_trait::async_trait;
use reqwest;
use serde_json;
use std::collections::HashMap;
use crate::plugins::traits::{LLMPlugin, PluginError};
use crate::plugins::llm::mock::{LLMRequest, LLMResponse};

/// Hugging Face plugin for open source models
#[cfg(feature = "llm-apis")]
pub struct HuggingFaceLLMPlugin {
    api_key: String,
    model: String,
    client: reqwest::Client,
}

#[cfg(feature = "llm-apis")]
impl HuggingFaceLLMPlugin {
    pub fn new(api_key: String, model: String) -> Self {
        Self {
            api_key,
            model,
            client: reqwest::Client::new(),
        }
    }
}

#[cfg(feature = "llm-apis")]
#[async_trait]
impl LLMPlugin for HuggingFaceLLMPlugin {
    async fn generate(&self, request: &LLMRequest) -> Result<LLMResponse, PluginError> {
        let mut payload = HashMap::new();
        payload.insert("inputs", request.prompt.clone());

        let mut parameters = HashMap::new();
        if let Some(max_tokens) = request.max_tokens {
            parameters.insert("max_new_tokens", max_tokens);
        }
        if let Some(temperature) = request.temperature {
            parameters.insert("temperature", temperature as u32);
        }
        payload.insert("parameters", serde_json::to_string(&parameters)?);

        let url = format!("https://api-inference.huggingface.co/models/{}", self.model);

        let response = self
            .client
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("Content-Type", "application/json")
            .json(&payload)
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(PluginError::Api(format!(
                "Hugging Face API returned status: {}",
                response.status()
            )));
        }

        let response_data: serde_json::Value = response.json().await?;

        let content = if response_data.is_array() {
            response_data[0]["generated_text"]
                .as_str()
                .unwrap_or("No response generated")
                .to_string()
        } else {
            response_data["generated_text"]
                .as_str()
                .unwrap_or("No response generated")
                .to_string()
        };

        Ok(LLMResponse {
            content,
            tokens_used: None, // HF doesn't always provide token counts
            model: self.model.clone(),
        })
    }

    fn get_model_name(&self) -> String {
        self.model.clone()
    }
}
